include/godot_cpp/classes/wrapped.hpp

```diff
@@ -41,7 +41,9 @@
 #include <godot_cpp/godot.hpp>
 
 namespace godot {
-
+namespace internal {
+_ALWAYS_INLINE_ void _set_construct_info(const StringName *p_extension_class_name, const GDExtensionInstanceBindingCallbacks *p_binding_callbacks);
```

Comment:
Because it should be a **public** method to be called by `_pre_initialize`, if it become a static method of `Wrapped`, it may be called by users. So I think `godot::internal` is a good place of this method.

Reply:
Could we make `_pre_initialize()` a friend of `Wrapped`, allowing the static method to be private or protected?

Reply:
I had struggled with this, but failed at last. I don't quite remember the details anymore, I only know the reason is about template function and friend definition.

Reply:
I was able to get this working for me, at least with GCC. Here's the patch:

```patch
diff --git a/include/godot_cpp/classes/wrapped.hpp b/include/godot_cpp/classes/wrapped.hpp
index 601dd61..dde0573 100644
--- a/include/godot_cpp/classes/wrapped.hpp
+++ b/include/godot_cpp/classes/wrapped.hpp
@@ -41,24 +41,30 @@
 #include <godot_cpp/godot.hpp>
 
 namespace godot {
-namespace internal {
-_ALWAYS_INLINE_ void _set_construct_info(const StringName *p_extension_class_name, const GDExtensionInstanceBindingCallbacks *p_binding_callbacks);
-} //namespace internal
 class ClassDB;
 
 typedef void GodotObject;
 
+template <typename T, std::enable_if_t<std::is_base_of<::godot::Wrapped, T>::value, bool> = true>
+_ALWAYS_INLINE_ void _pre_initialize();
+
 // Base for all engine classes, to contain the pointer to the engine instance.
 class Wrapped {
        friend class GDExtensionBinding;
        friend class ClassDB;
        friend void postinitialize_handler(Wrapped *);
 
-       friend _ALWAYS_INLINE_ void internal::_set_construct_info(const StringName *p_extension_class_name, const GDExtensionInstanceBindingCallbacks *p_binding_callbacks);
+       template <typename T, std::enable_if_t<std::is_base_of<::godot::Wrapped, T>::value, bool>>
+       friend _ALWAYS_INLINE_ void _pre_initialize();
 
        thread_local static const StringName *_constructing_extension_class_name;
        thread_local static const GDExtensionInstanceBindingCallbacks *_constructing_class_binding_callbacks;
 
+       _ALWAYS_INLINE_ static void _set_construct_info(const StringName *p_extension_class_name, const GDExtensionInstanceBindingCallbacks *p_binding_callbacks) {
+               _constructing_extension_class_name = p_extension_class_name;
+               _constructing_class_binding_callbacks = p_binding_callbacks;
+       }
+
 protected:
        bool has_object_instance_binding() const;
        virtual bool _is_extension_class() const { return false; }
@@ -117,16 +123,9 @@ public:
        GodotObject *_owner = nullptr;
 };
 
-namespace internal {
-_ALWAYS_INLINE_ void _set_construct_info(const StringName *p_extension_class_name, const GDExtensionInstanceBindingCallbacks *p_binding_callbacks) {
-       Wrapped::_constructing_extension_class_name = p_extension_class_name;
-       Wrapped::_constructing_class_binding_callbacks = p_binding_callbacks;
-}
-} //namespace internal
-
 template <typename T, std::enable_if_t<std::is_base_of<::godot::Wrapped, T>::value, bool> = true>
 _ALWAYS_INLINE_ void _pre_initialize() {
-       internal::_set_construct_info(T::_get_extension_class_name(), &T::_gde_binding_callbacks);
+       Wrapped::_set_construct_info(T::_get_extension_class_name(), &T::_gde_binding_callbacks);
 }
 
 _FORCE_INLINE_ void snarray_add_str(Vector<StringName> &arr) {
```

If you apply that to this PR, then CI will run, and we can see if it works with clang and MSVC as well.

Reply:
It works with clang and MSVC, but without GCC.

Reply:
My patched version works for me in GCC!

Comparing the code in your PR to my patch above, it looks like you didn't include the forward declaration? Try adding this code above `class Wrapped {`:

```c++
template <typename T, std::enable_if_t<std::is_base_of<::godot::Wrapped, T>::value, bool> = true>
_ALWAYS_INLINE_ void _pre_initialize();
```

Reply:
Oh, it is my mistake. Already fixed.

Reply:
Thanks! Looks like it's working :-)